pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -571,6 +596,145 @@ func (a Adapter) generateDeploymentObjectMeta(deployment *appsv1.Deployment, lab
+	}
+
+	isSBOSupported, err := a.kubeClient.IsServiceBindingSupported()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error in determining if the Service Binding Opertor is supported, cause: %w", err)
+	}
+
+	for _, remoteK := range remoteK8sResources {
+		matchFound := false
+		for _, devfileK := range devfileK8sResourcesUnstructured {
+			// only check against GroupKind because version might not always match
+			remoteResourceIsPresentInDevfile := devfileK.GroupVersionKind().GroupKind() == remoteK.GroupVersionKind().GroupKind() &&
+				devfileK.GetName() == remoteK.GetName() &&
+				remoteK.GetNamespace() == currentNamespace
+			if remoteResourceIsPresentInDevfile {
+				matchFound = true
+				break
+			}
+
+			// check for secrets related to ServiceBinding
+			// if the SBO is installed, we can mark the secret as match if its ServiceBinding resource is found locally;
+			// we do this because such secrets will be owned by the ServiceBinding resource and will be handled by it
+			// if the SBO is not installed, we check if the secret's labels matches those created by odo for binding
+			if remoteK.GroupVersionKind() == kclient.SecretGVK {
+				var remoteSecretHasLocalServiceBindingOwner bool
+				if isSBOSupported {
+					// if SBO is installed, check if its owner reference is a servicebinding present in the devfile;
+					// if the owner matches, consider it a match found
+					remoteSecretHasLocalServiceBindingOwner = func() bool {
+						for _, owner := range remoteK.GetOwnerReferences() {
+							if owner.Kind == kclient.ServiceBindingKind && owner.Name == devfileK.GetName() {
+								return true
+							}
+						}
+						return false
+					}()
+
+				} else {
+					// if SBO is not installed, check if it is a servicebinding secret
+					remoteSecretHasLocalServiceBindingOwner = service.IsLinkSecret(remoteK.GetLabels()) &&
+						remoteK.GetLabels()[service.LinkLabel] == devfileK.GetName() &&
+						remoteK.GetNamespace() == currentNamespace
+				}
+				if remoteSecretHasLocalServiceBindingOwner {
+					matchFound = true
+					break
+				}
+			}
+		}
+		if !matchFound {
+			// only remove servicebinding secrets if the SBO is not installed
+			if !isSBOSupported && remoteK.GroupVersionKind() == kclient.SecretGVK && service.IsLinkSecret(remoteK.GetLabels()) {
+				serviceBindingSecretsToRemove = append(serviceBindingSecretsToRemove, remoteK)
+			} else {
+				objectsToRemove = append(objectsToRemove, remoteK)
+			}
+		}
+	}
+	return objectsToRemove, serviceBindingSecretsToRemove, nil
+}
+
+// deleteRemoteResources takes a list of remote resources to be deleted
+func (a Adapter) deleteRemoteResources(objectsToRemove []unstructured.Unstructured) error {
+	// Delete the resources present on the cluster but not in the Devfile
+	for _, objectToRemove := range objectsToRemove {
+		spinner := log.Spinnerf("Deleting Kubernetes resource: %s/%s", objectToRemove.GetKind(), objectToRemove.GetName())
```

Comment:
Safer to call `defer spinner.End(false)` right after this. Otherwise, spinner will never end if `gvr, err := a.kubeClient.GetGVRFromGVK(objectToRemove.GroupVersionKind())` line below returns a non-nil error.

Reply:
You're right. I missed that.